### ISSUE TITLE
[Search.gov] Rate limiting errors for the /v0/search_click_tracking endpoint that reports analytics for typeahead

### ISSUE THREAD

**Issue by jilladams** [department-of-veterans-affairs/va.gov-cms#14819]
## Description

We are seeing errors related to Search.gov, but it's not clear if those errors are Veteran facing. 
If so: this is urgent to ticket / address
If not: this is low priority

### User story

**AS A** Veteran
**I WANT** to not see errors
**SO THAT** I can successfully search

**AS A** Platform engineer
**I WANT** errors to be less noisy / more meaningful
**SO THAT**  I can effectively triage emergencies

### Engineering notes / background
Thread: https://dsva.slack.com/archives/CBU0KDSB1/p1690239844318679
> issue with `/v0/search_click_tracking`: If I'm reading things right, over half of requests to this endpoint have failed over the last week with a 499 status code from traefik (stet)

~71k 204s: [https://vagov.ddog-gov.com/logs?query=%22POST%20%2F%2Fv0%2Fsearch_click_tracking%22%20[…]=stream&from_ts=1689634744454&to_ts=1690239544454&live=true](https://vagov.ddog-gov.com/logs?query=%22POST%20%2F%2Fv0%2Fsearch_click_tracking%22%20%221.1%5C%22%20204%22%20&cols=host%2Cservice&index=%2A&messageDisplay=inline&stream_sort=desc&viz=stream&from_ts=1689634744454&to_ts=1690239544454&live=true)

~41k 499s: [https://vagov.ddog-gov.com/logs?query=%22POST%20%2F%2Fv0%2Fsearch_click_tracking%22%20[…]=stream&from_ts=1689634231185&to_ts=1690239031185&live=true](https://vagov.ddog-gov.com/logs?query=%22POST%20%2F%2Fv0%2Fsearch_click_tracking%22%20%221.1%5C%22%20499%22%20&cols=host%2Cservice&event=AgAAAYmKHP58wPHaNQAAAAAAAAAYAAAAAEFZbUtIUTRVQUFDck83Zm5UbTZNYXdCUgAAACQAAAAAMDE4OThhMWQtNGJkYS00MTM5LWJjNzktMmQyMWJkZTBhNGIy&index=%2A&messageDisplay=inline&stream_sort=desc&viz=stream&from_ts=1689634231185&to_ts=1690239031185&live=true)

**Docs:** All known docs about how Search.gov / on-site search work: https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/products/on-site-search

## Acceptance criteria

- [ ] Determine if the reported errors are veteran facing or not, ticket next steps
   - [ ] If so: pull into sprint, raise to PM/DM for awareness

**Comment by wesrowe** [comment]
I'm racking my brain for other avenues we could try.

- we can look at search.gov analytics once our accounts are reactivated
- we could also try opening a support ticket with search.gov, if we can determine that the vets-api endpoint is passing requests through to search.gov

@randimays, did you already look at vets-api code for clues?

**Comment by randimays** [comment]
@wesrowe I did some digging in vets-api, but I'm not Ruby-literate so I may have to bubble some questions up to the team.

Planning to get to the Search.gov analytics shortly. Had to close the loop on other tickets.

**Comment by randimays** [comment]
Notes from research (so far):

- `on-site-search` documentation is lacking in spots. See [click tracking README](https://github.com/department-of-veterans-affairs/va.gov-team/blob/3e0b1c9970e17b43cdf8e503be87ebb84c97148d/products/on-site-search/click-tracking/readme.md) and [docs for analytics](https://github.com/department-of-veterans-affairs/va.gov-team/tree/master/products/on-site-search/analytics). Maybe we have some opportunities here?

- The [search playbook](https://github.com/department-of-veterans-affairs/vets-website/pull/25237) confirms vets-api is sending this data to Search.gov for improving search algorithms and typeahead
  - [vets-api code where the Search.gov click tracking API is wrapped](https://github.com/department-of-veterans-affairs/vets-api/blob/master/lib/search_click_tracking/service.rb)
  - [vets-api code where the API behavior is defined](https://github.com/department-of-veterans-affairs/vets-api/blob/master/app/controllers/v0/search_click_tracking_controller.rb)

- Confirmed using a search query of "bananas" that the data is indeed posting to the [Search.gov analytics dashboard](https://search.usa.gov/sites/7378/queries/new).

<img width="1643" alt="Screenshot 2023-08-23 at 2 49 33 PM" src="https://github.com/department-of-veterans-affairs/va.gov-cms/assets/19175324/fbf96bda-334b-4735-87f1-8cb9f880a477">

- vets-website calls the `search_click_tracking` endpoint [here](https://github.com/department-of-veterans-affairs/vets-website/blob/a5b465cb1478652a819aa0c43025caaa55685740/src/applications/search/containers/SearchApp.jsx#L179C4-L179C4) 

I opened a support ticket with Search.gov (via email). I mentioned that we were having issues with this endpoint for quite some time, but didn't provide any detail. This was the support person's reply

> Click data is used to improve [search.gov](http://search.gov/) results- more commonly clicked results are ranked higher in subsequent searches.
> Could you please outline the issues that you have been seeing with VA click data so that we can take a closer look?

Checking with @wesrowe to see what kind of details I can share with this group. Not sure if we can hand over logs or screenshots to logs, or whether a simple text-only outline of the problem will suffice.

**Comment by randimays** [comment]
Update: I sent a reply on the support ticket with Search.gov. Adrian has been tremendously helpful and made some suggestions for what to ask them about ([see this thread](https://dsva.slack.com/archives/C52CL1PKQ/p1692883308106199)). I sent that late Wednesday and haven't heard back yet.

**Comment by randimays** [comment]
Update: I got a reply from Search.gov. They confirmed they have standard rate-limiting practices in place, but VA IP addresses are allow-listed. They sent me the IP addresses they have for VA across environments and asked me to verify their accuracy. I reached out to Platform to see if someone could help me with this.

Platform let me know that such a request needs to be directed to the NEO team (aka TIC Gateway Operations) that manages the public DNS or devices used for traffic in and out of the VA network.

Here is their contact information:
```
Phone: (304) 262-5282 (sometimes goes directly to voicemail)
Email: nocnetworkedgeoperations@va.gov
ServiceNow group: IO.NETWORK.NOC.NEO
```

I have reached out to them with the request.

**Comment by randimays** [comment]
Search.gov support folks have discovered:

> It looks like your requests are being caught in one of the automated security filters that we have set up.

They do need us to confirm the IP addresses they sent over last week. My SNOW ticket doesn't have a response yet. I pinged on it and may call them tomorrow if there's no movement.

CC: @acrollet

**Comment by acrollet** [comment]
oooh, I actually just found good docs on this:

https://github.com/department-of-veterans-affairs/devops/blob/master/docs/System-Locations.md#routing-through-externalpublic-network

and 

https://github.com/department-of-veterans-affairs/devops/blob/master/docs/VA%20Network%20Translation.md

It has been long enough since these have been updated (3 years) that we should still confirm the data on these docs with NOC NEO and either update them or make a note of the date confirmed.

**Comment by acrollet** [comment]
thank you!

**Comment by randimays** [comment]
ESD has begun responding; they asked for further clarification of the issue that is happening, and a reference ticket (CID# or RFC#). They also copied a colleague that should be able to verify the IP address information. I sent them an explanation of the issue this morning.

**Comment by randimays** [comment]
[Slack thread](https://dsva.slack.com/archives/C52CL1PKQ/p1694103535286779) starting with CID or RFC. Adrian pulled in MikeC also.

**Comment by mchelen-gov** [comment]
SNOW issue: [INC28626006](https://yourit.va.gov/now/nav/ui/classic/params/target/incident.do%3Fsys_id%3D2181c73687b83594e99f77f5cebb3565%26sysparm_view%3D)

**Comment by randimays** [comment]
We got confirmation from ESD that the IP addresses Search.gov provided us with are accurate. I sent the word to the Search.gov folks; waiting on their reply.

**Comment by randimays** [comment]
Search.gov replied. They are going to keep the IP addresses on the allow list. They made a change in their firewall settings last week to see if it helped with the rate of errors. There wasn't any measurable change, unfortunately. @mchelen-gov and @acrollet had some questions for the support team, so I reiterated those in my reply:

> - Can you see what other IP addresses are making requests with our API key?
> - Can we remove the IP allow-listing, even if we had to move to a more frequent API key rotation schedule?
> - Are you able to do IP address wildcards? The full IP may change. Such as:
> 
>  152.130.*.*
>  152.131.*.*
>  152.132.*.*
>  152.133.*.*

**Comment by randimays** [comment]
We're finished communicating with the Enterprise Service Desk, so I closed the ticket with them.

Search.gov is still looking into the above questions. They found a rate-limiting parameter that they are going to update to support VA specifically.

They also said that the calls we shared (`/v0/search_click_tracking`) are translated to the [GSA](open.gsa.gov) click tracking endpoint. Those begin with `api.gsa.gov`. They asked for some examples of failing calls for that. Adrian helped me with spelunking and I sent some examples over and am waiting for their reply.

**Comment by randimays** [comment]
The last update we had from the Search.gov support team was last Friday afternoon. I sent a follow-up to my contact just now.

**Comment by randimays** [comment]
The latest from Search.gov:

> We think that there are two potential issues going on.  There is some amount of rate-limiting, which is the source of some of the errors that your are seeing.  This is a specific issue that we think we have a solution for. 
> 
>It also appears that there is also some amount of data that is just not making it from one part of the search system where clicks are actually recorded, and another part, where you see the roll-up of data, and a developer is digging into this issue right now.

**Comment by randimays** [comment]
End of sprint update: everything is the same as above ^

**Comment by randimays** [comment]
This was the latest status-check reply from Search.gov on October 4th:

> We're looking at adjusting the rate limit. Going back to the 269k requests you sent that week in August, do you have a sense for the peak hourly number you see? And do you know if that week was typical? What was the number from last week? 

I sent them replies to their questions and haven't yet heard back. This will carry over to next sprint.

**Comment by randimays** [comment]
Latest status-check reply from Search.gov on October 11th:

> I would be comfortable adjusting our rate limit to 5500 per hour. This averages to about 1.5 requests per second, which I feel is reasonable. Do you use a load balancer to distribute requests across your IPs? If so, would it be fair to say that for each of the 6 IPs we know about, they send less than 1k each hour?

I have a question out to Mike Chelen and to the NOC NEO team to get some answers on the above questions.

**Comment by randimays** [comment]
NOC NEO team (via email) told me to open another ServiceNOW ticket. I opened RITM10346501, but that's likely the wrong category of ticket so I'll wait for someone to tell me where to put it.

**Comment by acrollet** [comment]
@randimays thanks for sticking with this difficult process!

**Comment by randimays** [comment]
Adrian brought up some great points regarding the proposed rate limit adjustment to 5500, and the question about requests from individual IPs. [Slack thread](https://dsva.slack.com/archives/C52CL1PKQ/p1697150309403499?thread_ts=1692883308.106199&cid=C52CL1PKQ)

I asked Search.gov if there is a ceiling on the rate limit because we are anticipating higher traffic soon, and if they could see the requests per hour by IP in their logs. Thanks @acrollet!

**Comment by randimays** [comment]
Latest reply from Search.gov:

> I found a way to pull numbers for the individual IPs that we knew about for VA. It turned out none of these were sending requests, going all the way back to July 1. We DO show 1,895,861 requests from the Faraday v0.17.6 user agent, though, and I can see that useragent on the successful click requests that come through to our system from you (no other customers use this agent). These requests come from:
> 
>52.222.123.172
>52.222.122.97
> 
>I'm asking the team that manages the API analytics how I can see more detail. One question they had was if you could provide the IP address(es) that is(are) giving you the 429s?

I dug around in the Datadog logs for the known VA prod IPs and was not able to find them in the 499s we receive from Search.gov. Hoping @acrollet or @mchelen-gov can help?

**Comment by randimays** [comment]
Latest reply from Search.gov:

>  I was able to get to the right place in the analytics, and over the past six weeks we've seen basically two IPs handling your requests - they are not perfectly balanced, but it's close to 50/50 (this past week 55/45):
 >  
>  152.129.43.1
>  152.129.43.2
 >  
>  Interesting that these were not on our radar before, but now they are, which is good.
>  
>  It looks like things are pretty much successful at this point, with 95k getting a 200 response (this matches our internal logs), and about 7k getting a 429.
 >  
>  The peak number of requests we saw from the two IPs combined in an hour was just shy of 2200, so I'm going to boost the limit to 1100 per IP for now, and hopefully that will resolve things for the time being.

It seems problematic for them to raise the rate limit to barely above what the peak is **now**, considering we mentioned expecting our traffic to increase. I'll see if we can get them to raise the limit higher so we don't have to go through this again in 6 months (or however long it takes for the new health portal on VA.gov to take shape).

**Comment by acrollet** [comment]
> It seems problematic for them to raise the rate limit to barely above what the peak is **now**, considering we mentioned expecting our traffic to increase. I'll see if we can get them to raise the limit higher so we don't have to go through this again in 6 months (or however long it takes for the new health portal on VA.gov to take shape).

💯 💯 💯 

thank you! I would ask for at least 3x if not more, we'll expect some big traffic surges when communications go out to Veterans.

**Comment by randimays** [comment]
I sent a message to Search.gov 2 days ago asking if they could increase our rate limit per IP to 4,000 or 5,000 per hour. Here is the response:

> as a counter proposal, to keep the potential for DDOS from other actors to a minimum...
> 
>Would it be possible to add hosts to the set that send the requests to us? For instance, we know about 8 unique IPs for y'all, but currently only 2 are sending the click requests to us.
> 
>and
> 
>What date is the health portal going live?

I'm getting out of my depth here - not sure who or where to go to get these questions answered. @jilladams any ideas?

**Comment by acrollet** [comment]
The health portal will be rolling out over the next few months and parts of it are already live for beta users. TBH I really find this concern overblown on their part. To me it would be ideal if they rate limited on the API key (+ a netmask if they absolutely must) if possible.

**Comment by randimays** [comment]
@acrollet Really appreciate you leaning in here to assist! I asked this question; hopefully they can help us out here.

**Comment by randimays** [comment]
I was out on PTO last week.

I sent this question on Oct 18:
> The health portal will be rolling out over the next few months. Parts of it are already live for beta users.
>
> Are you able to rate limit based on the API key (and a netmask if necessary)?

This is Search.gov's reply:
> it's a bit complicated to implement a key-based limit, because we provide our own keys, rather than use keys issued by the API umbrella that manages the rate limits. 
>
> Do I perceive correctly that it would not be easy to let more IP addresses participate in sending requests to us?

Given that I'm not sure if this is possible and it'd likely take weeks of a SNOW request to get an answer, I'm thinking we should tell them this is quite complex, if possible at all. @acrollet Any suggestions on how to proceed?

**Comment by acrollet** [comment]
⚠️ hand waving alert ⚠️ 

My guess is that there are 8 IPs because there are 2 per TIC Gateway, so it actually depends on which TIC VA.gov traffic is going through. Short answer is I find it highly unlikely that we could get more addresses participating, but any of the 8 could send traffic at any time depending on configuration and failover events.

**Comment by randimays** [comment]
Thank you @acrollet. By the way, if you have / find any resources to learn about this complex networking stuff that you think might be helpful for me, please send them my way!

**Comment by randimays** [comment]
This was the last reply I sent to Search.gov on Oct 30:
>We think there are 8 IPs because there are 2 IPs per TIC gateway. It actually depends on which TIC [va.gov](http://va.gov/) traffic is going through. We find it highly unlikely that we could get more IP addresses participating. But any of the 8 could send traffic at any time depending on configuration and failover events.

Waiting on a response from them. This ticket will likely carry over as it seems like we are having to negotiate here.

**Comment by randimays** [comment]
No reply from Search.gov yet. I sent a followup email to see if they're able to raise our rate limit to 4,000 or 5,000 per IP address. They seem hesitant about it but I'll keep trying.

**Comment by randimays** [comment]
Handed off this ticket and the very long email thread (it makes Gmail almost crash for me) to @jilladams and @FranECross. I'll stand by as support though.

**Comment by acrollet** [comment]
Just to say it, if we're not using the data for our own purposes (unknown by me) and search.gov isn't interested enough to enable us to send it all, maybe we disable this or limit to a percentage of users? (This would likely be a Dave or Danielle decision)

**Comment by jilladams** [comment]
Summarizing from comments thread, just for the sake of knowing what happened / where we are: 

## The issue
A monitoring event revealed that some % of calls to `/v0/search_click_tracking`  (analytics for Search.gov) fail with 429 status code from Traefik.
  * Veteran impact is negligible.
  * What this endpoint call does: vets-api sends data to Search.gov to improve search algorithms and typeahead. ("Click data is used to improve [search.gov](http://search.gov/) results- more commonly clicked results are ranked higher in subsequent searches.")
    * [vets-api code where the Search.gov click tracking API is wrapped](https://github.com/department-of-veterans-affairs/vets-api/blob/master/lib/search_click_tracking/service.rb)
    * [vets-api code where the API behavior is defined](https://github.com/department-of-veterans-affairs/vets-api/blob/master/app/controllers/v0/search_click_tracking_controller.rb)

## Troubleshooting 
so far has included: 
* **IP addresses:** Enterprise Service Desk verified our IP addresses ([INC28626006](https://yourit.va.gov/now/nav/ui/classic/params/target/incident.do%3Fsys_id%3D2181c73687b83594e99f77f5cebb3565%26sysparm_view%3D)), and Search.gov verified those IP addresses are allow-listed
  * We have 8 specified IP addresses. Only 2 send traffic to this endpoint: 152.129.43.1 and 152.129.43.2
  * We believe the 8 IPs are 2 IPs per TIC gateway, and that routing from va.gov > TIC determines which IP addresses get used.
  * Any of the 8 IPs could send traffic at any time depending on configuration and failover events, but we do not think we can force distribute the traffic from these 2 active IPs across the other 6 available IPs. 

* **Rate limits:** Search.gov has rate limiting in place. 
  * When the 2 active IPs were identified, Search.gov set our limit to 1100 / hour per IP, based on peak traffic from the August week where this monitor went off.
  * We would like it raised to higher than peak, to be prepared for traffic increases when the health portal launches. 
  * Search.gov instead requested we redistribute our load across additional IPs, which we don't think we can do bc the TIC note above.
  * Our ideal rate limit would be ~5500 per hour total (~1.5 requests per second average), rather than 2200/hour.
    * Search.gov doesn't want to increase the rate limit for our 2 IPs to that high a rate, out of concern for potential DDOS activity. 
  * **API key based rate limits**: We requested this, but Search.gov responded that's not easy to implement because they "provide our own keys, rather than use keys issued by the API umbrella that manages the rate limits."
  * Right now: Stalemate. 
* **Tracing our calls:**
  * Our requests were initially being caught in an automated security filter.
  * `/v0/search_click_tracking` calls are translated to the [GSA](https://github.com/department-of-veterans-affairs/va.gov-cms/issues/open.gsa.gov) click tracking endpoint, beginning with `api.gsa.gov`. Some amount of data is not making it from one part of the search system where clicks are actually recorded, to another part, where you see the roll-up of data. Search.gov is investigating. No updates on this front.

Other questions we've asked, not answered while folks at Search.gov pursued other pieces of the thread: 
  * Can you see what other IP addresses are making requests with our API key?
  * Can we remove the IP allow-listing, even if we had to move to a more frequent API key rotation schedule?
  * Are you able to do IP address wildcards? The full IP may change. Such as: 
    * 152.130..
    * 152.131..
    * 152.132..
    * 152.133..

For now: I've followed up via email with Search.gov to ask about the rate limit increase again, to see what they say.

**Comment by jilladams** [comment]
Search.gov let me know this week that the cc/bcc change affected their support interface, hence the delayed response. Latest note asked: 

> - Could you provide an update on how many IP addresses you plan to add to your load balancer, and if you're still getting errors? The current setting is at 1100 per IP, but perhaps the increase in reporting/calls hasn't begun yet? Since this API tracks activity on clicks, I think the likelihood of DOS attempts is pretty low (knock wood!), so I would be comfortable experimenting with an increase when you are adding an IP to your load balancer.
> - We have a few layers of DOS protection in place - one is managed by the allow-list you mentioned that I mentioned. When you add IPs, we will want to know what they are, so we can add them to this list. The other is managed by the rate limit in our API umbrella ([api.data.gov](http://api.data.gov/) is the backing service, if you're familiar with that).


I've replied: 
> RE: load balancing: Any of our 8 IPs could send traffic at any time depending on configuration and failover events, but we do not think we can force distribute the traffic from these 2 active IPs across the other 6 available IPs. We don't currently have intentions to add more IPs. 

> We have seen additional issues this week. On both Jan 18 and Jan 26, we got 500 and 504 errors from calls to [search.gov](http://search.gov/), with a message body that said: h2>This website is under heavy load (queue full)</h2><p>We're sorry, too many people are accessing this website at the same time. We're working on this problem. Please try again later.</p>

> Please let me know if you are comfortable increasing our API limits, with that in mind. Thanks!

**Comment by jilladams** [comment]
Regarding the unknown error message, I did a little digging around -- it looks like this is a default message used by Phusion passenger: 
https://stackoverflow.com/questions/31670217/this-website-is-under-heavy-load-rails-app
https://www.phusionpassenger.com/library/config/nginx/reference/#passenger_max_request_queue_size

We didn't find anywhere we use that in the stack: https://dsva.slack.com/archives/C05THHJHH2R/p1707151355628999?thread_ts=1706564391.025529&cid=C05THHJHH2R

From search.gov: 
> Hi Jill, so it turns out we do use Phusion Passenger, and also that we issue out-of-the-box errors to API requests, and not our custom errors (because API connections don't like pretty pages). That said, looking at the traffic patterns for the two windows you mentioned, we were not under heavy load at that time, and we weren't doing any maintenance, so it doesn't make sense that we would have been handing out errors.

I don't think we are going to get any traction on this without escalating somehow to ask for better error handling on their end.
